rusoto/core/src/request.rs

```diff
@@ -77,19 +78,61 @@ impl From<IoError> for HttpDispatchError {
 
 ///Trait for implementing HTTP Request/Response
 pub trait DispatchSignedRequest {
+    /// TODO: docs
+    type Future: Future<Item=HttpResponse, Error=HttpDispatchError> + 'static;
     /// Dispatch Request, and then return a Response
-    fn dispatch(&self, request: &SignedRequest) -> Result<HttpResponse, HttpDispatchError>;
+    fn dispatch(&self, request: SignedRequest) -> Self::Future;
 }
 
-impl DispatchSignedRequest for Client {
-    fn dispatch(&self, request: &SignedRequest) -> Result<HttpResponse, HttpDispatchError> {
+/// TODO: docs
+pub struct ClientFuture(ClientFutureInner);
+
+enum ClientFutureInner {
+    Hyper(HyperFutureResponse),
+    Error(String)
+}
+
+impl Future for ClientFuture {
+    type Item = HttpResponse;
+    type Error = HttpDispatchError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
```

Comment:
So I don't see support for timeouts here. This is something that I'd really love to have (and we do it in the credentials crate). Am I missing it, or is it not a part of it?

Reply:
Nope, it's not in here yet. To me this seems like something we could ship after the initial async changes, since a) we can very likely add it later w/o breaking the API and b) it's not part of the current rusoto feature set either. What do you think?

Reply:
I think we'd want at least some basic support for timeouts. I know it's not part of the feature set, but I could very well see someone using the async provider, and wondered why certain code isn't getting called. That extra layer of indirection I think makes it harder to easily spot "oh it's a timeout"

Plus personally I think it's much easier to support timeouts with futures.

Reply:
Per parent comments I agree we want timeouts, but if adding them to this PR makes it a lot larger it's harder to get the changes into `master`.  I'm okay with leaving them out of this PR and getting them in as soon as this one lands. 😄

Reply:
Adding timeouts shouldn't add too much to this commit which is why I suggest adding it in this one (in previous it was a couple lines of error handling). Tokio can natively support this by joining two futures together (this is what the creds crate does now).